agent/hcp/telemetry_provider.go

```diff
@@ -0,0 +1,191 @@
+	labels   map[string]string
+	filters  *regexp.Regexp
+	// refreshInterval controls the interval at which configuration is fetched from HCP to refresh config.
+	refreshInterval time.Duration
+}
+
+// providerParams is used to initialize a hcpProviderImpl.
+type providerParams struct {
+	metricsConfig   *client.MetricsConfig
+	refreshInterval time.Duration
+	hcpClient       client.Client
+}
+
+// NewHCPProviderImpl initializes and starts a HCP Telemetry provider with provided params.
+func NewHCPProviderImpl(ctx context.Context, params *providerParams) (*hcpProviderImpl, error) {
+	if params.hcpClient == nil {
+		return nil, errors.New("missing HCP client")
+	}
+
+	if params.metricsConfig == nil {
+		return nil, errors.New("missing metrics config")
+	}
+
+	if params.refreshInterval <= 0 {
+		return nil, fmt.Errorf("invalid refresh interval: %d", params.refreshInterval)
+	}
+
+	cfg := &dynamicConfig{
+		endpoint:        params.metricsConfig.Endpoint,
+		labels:          params.metricsConfig.Labels,
+		filters:         params.metricsConfig.Filters,
+		refreshInterval: params.refreshInterval,
+	}
+
+	hash, err := calculateHash(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate hash: %w", err)
+	}
+
+	t := &hcpProviderImpl{
+		cfg:       cfg,
+		cfgHash:   hash,
+		hcpClient: params.hcpClient,
+	}
+
+	go t.run(ctx)
+
+	return t, nil
+}
+
+// run continously checks for updates to the telemetry configuration by making a request to HCP.
+// Modification of config only occurs if changes are detected to decrease write locks that block read locks.
+func (t *hcpProviderImpl) run(ctx context.Context) {
+	ticker := time.NewTicker(t.cfg.refreshInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if newCfg, newHash, hasChanged := t.checkUpdate(ctx); hasChanged {
+				t.modifyTelemetryConfig(newCfg, newHash)
+				ticker.Reset(newCfg.refreshInterval)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// checkUpdate makes a HTTP request to HCP to return a new metrics configuration and true, if config changed.
+// checkUpdate does not update the metricsConfig field to prevent acquiring the write lock unnecessarily.
+func (t *hcpProviderImpl) checkUpdate(ctx context.Context) (*dynamicConfig, uint64, bool) {
+	logger := hclog.FromContext(ctx).Named("telemetry_config_provider")
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	telemetryCfg, err := t.hcpClient.FetchTelemetryConfig(ctx)
+	if err != nil {
+		logger.Error("failed to fetch telemetry config from HCP", "error", err)
+		metrics.IncrCounter(internalMetricRefreshFailure, 1)
+		return nil, 0, false
+	}
+
+	newDynamicConfig := &dynamicConfig{
+		filters:         telemetryCfg.MetricsConfig.Filters,
+		endpoint:        telemetryCfg.MetricsConfig.Endpoint,
+		labels:          telemetryCfg.MetricsConfig.Labels,
+		refreshInterval: telemetryCfg.RefreshConfig.RefreshInterval,
+	}
+
+	newHash, err := calculateHash(newDynamicConfig)
```

Comment:
we only rarely check the hash (after every network request, which is infrequent). What do you think about, instead of returning `hash` from this func, and storing it on the `hcpProviderImpl`, we instead add a new method `equals` to dynamicConfig like:

```
func (d *dynamicConfig) equals(other *dynamicConfig) bool {
  return hashstructure.Hash(*d, hashstructure.FormatV2, nil) == hashstructure.Hash(*other, hashstructure.FormatV2, nil)
}
```

Small change, but would remove the `hash vars` and the `cfgHash` field